java/src/main/native/src/row_conversion.cu

```diff
@@ -447,32 +1073,558 @@ static inline int32_t compute_fixed_width_layout(std::vector<cudf::data_type> co
+                 [](auto c) { return c.null_mask(); });
+
+  auto dev_input_data = make_device_uvector_async(input_data, stream, mr);
+  auto dev_input_nm = make_device_uvector_async(input_nm, stream, mr);
+
+  std::vector<size_type> column_sizes;  // byte size of each column
+  std::vector<size_type> column_starts; // offset of column inside a row including alignment
+  column_sizes.reserve(num_columns);
+  column_starts.reserve(num_columns + 1); // we add a final offset for validity data start
+
+  auto schema_column_iter =
+      thrust::make_transform_iterator(thrust::make_counting_iterator(0),
+                                      [&tbl](auto i) -> std::tuple<data_type, column_view const> {
+                                        return std::make_tuple(tbl.column(i).type(), tbl.column(i));
+                                      });
+
+  size_type fixed_width_size_per_row = detail::compute_column_information(
+      schema_column_iter, schema_column_iter + num_columns, column_starts, column_sizes);
+
+  auto dev_col_sizes = make_device_uvector_async(column_sizes, stream, mr);
+  auto dev_col_starts = make_device_uvector_async(column_starts, stream, mr);
+
+  // total encoded row size. This includes fixed-width data, validity, and variable-width data.
+  auto row_size_iter = cudf::detail::make_counting_transform_iterator(
+      0, [fixed_width_size_per_row, num_columns] __device__(auto i) {
+        auto const bytes_needed =
+            fixed_width_size_per_row + util::div_rounding_up_safe<size_type>(num_columns, 8);
+        return detail::align_offset(bytes_needed, 8);
+      });
+
+  // fixed_width_size_per_row is the size of the fixed-width portion of a row. We need to then
+  // calculate the size of each row's variable-width data and validity as well.
+  auto validity_size = num_bitmask_words(num_columns) * 4;
+
+  auto batch_info = detail::build_batches(num_rows, row_size_iter, stream, mr);
+  auto gpu_batch_row_boundaries =
+      make_device_uvector_async(batch_info.batch_row_boundaries, stream);
+
+  // the first batch always exists unless we were sent an empty table
+  auto const first_batch_size = batch_info.row_batches[0].row_count;
+
+  std::vector<rmm::device_buffer> output_buffers;
+  std::vector<int8_t *> output_data;
+  output_data.reserve(batch_info.row_batches.size());
+  output_buffers.reserve(batch_info.row_batches.size());
+  std::transform(batch_info.row_batches.begin(), batch_info.row_batches.end(),
+                 std::back_inserter(output_buffers), [&](auto const &batch) {
+                   return rmm::device_buffer(batch.num_bytes, stream, mr);
+                 });
+  std::transform(output_buffers.begin(), output_buffers.end(), std::back_inserter(output_data),
+                 [](auto &buf) { return static_cast<int8_t *>(buf.data()); });
+
+  auto dev_output_data = make_device_uvector_async(output_data, stream, mr);
+
+  int info_count = 0;
+  detail::determine_windows(
```

Comment:
It seems like we're calling this function multiple times and having it compute the same thing.  Seems like it should be returning a vector of window information which then gets applied at the various later steps.

Reply:
We are trying to build the `block_info`, which is GPU data, for the table. This function is run once in order to properly size the array and then again to actually build the infos. It is non-trivial though, so it might be worth the effort to save off the windows and then use that array directly in building the block infos.

Reply:
Should I add an issue to track this for later or try to add it to this PR?